### ISSUE TITLE
What language is flake.nix written in?

### ISSUE THREAD

**Issue by dasJ** [NixOS/nix#4945]
This issue may sound silly but it's actually not clear to me what language they are written in.
The syntax and file ending both look like Nix but the [manual](https://nixos.org/manual/nix/stable/#chap-writing-nix-expressions) doesn't apply most of the time.
The manual makes it seem like this should work:
```nix
{
  description = "Testflake";
  outputs = import ./outputs.nix;
}
```
but running that results in
```
error: expected a function but got a thunk at /nix/store/lm4pyqnkbiwvp2kxjd64nqb2xkdxmbfx-source/flake.nix:4:3
```

So what language are flake.nix files written in (is there any specification or documentation) and why is that language not Nix?

**Comment by edolstra** [comment]
It's a subset of Nix that doesn't allow computation in the flake metadata attributes. So e.g. `outputs` cannot be a function application like `import ./outputs.nix`, it must be a function directly `outputs = { bla }: ...`. This is to prevent arbitrarily complex, possibly non-terminating computations while querying flake metadata.

This should be documented in the flake manpage.

**Comment by dasJ** [comment]
Will care about this when the feature gets remotely usable for my use case so please leave this open

**Comment by exarkun** [comment]
I ran into this and was quite surprised and confused for about half an hour.  I used `let ... in ...` at the top of my flake.nix and the resulting error message, "error: file '/nix/store/91d71n4sxkxnby3c3pbc8is4xm7a70r9-source/flake.nix' must be an attribute set" had me scratching my head (`builtins.typeOf` said it _evaluated_ to an attribute set) and looking all over for someplace I had put some non-set type into a set position.

I didn't believe the error message was worded accurately and precisely because I am used to Nix error messages that tell me one type was expected in place of another with little or no further location information and I thought this was one of those cases.

An error message something like "file '...' must be written using the limited flake expression language" might have helped me discover my mistake more quickly.

Also, none of the docs I read on my way to learning about flakes said anything about this limited expression language so apart from tripping over this error I'm not sure how I would have learned about this.

**Comment by dzmitry-lahoda** [comment]
could be awesome if there would be turing decidable(always halt) security attack protected(like infinite reference circular import parsing) nix which allows to be flake input ` limited flake expression language` and that to be documented as flake input.

**Comment by infinisil** [comment]
This issue has a better title and description, but it's a effectively a duplicate of https://github.com/NixOS/nix/issues/4945, which has a bunch more discussion.

**Comment by rhendric** [comment]
(@infinisil, did you mean #3966 maybe?)

**Comment by infinisil** [comment]
Ah yes thanks, fixed!

**Comment by matklad** [comment]
Let me attempt at a summary here. 

The problem to solve here is providing access to flake metadata as a plain inert data, which only needs parsing, but not evaluation. If this is _not_ done, than any operation with a flake requires evaluating nix code, which is slow and requires full `nix` tool. 

A negative example for this problem would be the Python ecosystem, where running setup.py is (used to?) required to learn about package dependencies, which could make dependency resolution very slow (as the resolver must sequentially evaluate the graph of transitive dependencies)

A positive example would be Rust, where the metadata is available in inert .toml and additionally mirrored in the registry, and, as a result, the tooling is pretty snappy. 

While inert metadata greatly speeds up and increases the reach of common use-cases, there are also uncommon use-cases which could make use of more computation power for metadata: [example](https://github.com/NixOS/nix/issues/3966#issuecomment-940788246)

Assuming we want to have inert metadata, there are two "dimension" for getting there:

* Split flake definition into two files, inert `flake.jsontomlwhatever` and `flake.nix` with code, OR use a single file, `flake.nix`
* Pick a language for inert subset, it could be JSON, TOML, or a data-only subset of nix

Status quo: a single file is used, with restricted subset of nix for metadata. 

TOML support was attempted at here: https://github.com/NixOS/nix/issues/3966#issuecomment-706734581

**Comment by matklad** [comment]
Notes from myself:

Zig is designing a similar system at the moment. The [current thinking](https://ziggit.dev/t/new-zig-build-system-documentation/2191/5) is to have `build.zig` with code to describe how to build the current package, and `build.zon` (zig object notation) with data to express external dependencies. That is, two files, but using a data-only subset of the language

---

Disregarding practical implications and thinking from the first principles, using data-only subset of JavaScript (JSON) feels inferior to using data-only subset of nix.

**Comment by balazs-lengyel** [comment]
It seems to me that the opposing sides of this problem are not easily unify-able. on one hand dynamic behavior is really nice in some cases but on the other hand not everyone should always pay this cost.

One solution could be template programming:
- keep `flake.lock` and `flake.nix` as is
- add optional `flake.template.nix` which is written in fully featured nix
- add a `nix flake template` sub-command to pre-calculate arbitrary nix expressions from `flake.template.nix` and (over-)write the result to `flake.nix`.

this way we could:
- have a standard format (`flake.nix`) for tooling
- arbitrary nix expressions and time complexity in template
- fast info for flake users
- an easy way to script automatic template->flake updates, if needed
- can make `flake.nix` even more strict if needed for performance and recommend moving `flake.nix` to `flake.template.nix` if nix becomes too strict about flake evaluation.

what we lose:
- one more file and sub-command

(got here from experiencing https://github.com/NixOS/nix/issues/3966 and https://github.com/NixOS/nix/issues/4384)

**Comment by Cloudef** [comment]
Tried to be clever and hit this with:
```nix
  inputs = let
    dep = url: { inherit url; inputs.nixpkgs.follows = "nixpkgs"; };
  in {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nix-darwin = (dep "github:LnL7/nix-darwin");
    home-manager = (dep "github:nix-community/home-manager");
    zig = (dep "github:Cloudef/nix-zig-stdenv");
    zls = (dep "github:zigtools/zls");
    hyprland = (dep "github:hyprwm/Hyprland");
    eww = (dep "github:elkowar/eww");
  };
```

**Comment by jorsn** [comment]
> One solution could be template programming:
>    
>  * keep `flake.lock` and `flake.nix` as is
> 
>  * add optional `flake.template.nix` which is written in fully featured nix
> 
>  * add a `nix flake template` sub-command to pre-calculate arbitrary nix expressions from `flake.template.nix` and (over-)write the result to `flake.nix`.
> 
>   -- from https://github.com/NixOS/nix/issues/4945#issuecomment-1864538239

This is already possible by running the more complicated `nix eval --raw -f flake.template.nix flake > flake.nix` instead of `nix flake template`, where `flake.template.nix` contains something similar to the code below. For projects where the inputs are so numerous or complex that actual code is really helpful, it might be worthwhile to take this approach.

```nix
# flake.template.nix
rec {
  description = "description";

  inputs = let
    dep = url: { inherit url; inputs.nixpkgs.follows = "nixpkgs"; };
  in {
    nixlib.url = "github:nix-community/nixpkgs.lib";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nix-darwin = (dep "github:LnL7/nix-darwin");
    home-manager = (dep "github:nix-community/home-manager");
    zig = (dep "github:Cloudef/nix-zig-stdenv");
    zls = (dep "github:zigtools/zls");
    hyprland = (dep "github:hyprwm/Hyprland");
    eww = (dep "github:elkowar/eww");
  };

  outputs = inputs: {
  };

  flake =
    let toPretty = import (builtins.fetchurl "https://gist.githubusercontent.com/jorsn/012be3e868736359024007fc87b631cf/raw/40bd0d0183215cb6b2cf5ac801559faabee78a0a/prettyUnsafe.nix");
    #or inherit (toPretty) (builtins.getFlake inputs.nixlib.url).lib.generators;
    #for less pretty but still correct output
    in ''
    {
      description = "${description}";

      inputs = ${toPretty "  " inputs};

      outputs = inputs: (import ./flake.tpl.nix).outputs inputs;
    }
  '';
}
```

For convenience, here is a link to the gist: https://gist.github.com/jorsn/012be3e868736359024007fc87b631cf

**Comment by roberth** [comment]
You could avoid some templating by treating `flake.nix` as mostly data, and then you put `outputs = inputs: import ./outputs.nix inputs;` at the end.

None of that should be needed though. We should consider actual solutions within flakes itself.

### Plain data inputs?

Regardless of where and how we represent it, the major use case for this seems to be `inputs`. Does that really need to be plain data?

- With the current style of locking, yes, but with #7730, no.
- As a way of determining all (possible) source inputs of a flake: not unless we also forbid `fetchTree` and fixed output derivations. As long as we have those, it's not feasible to figure out all actual inputs anyway, and I don't think we should forbid those functions. In other words, such static analysis is not _complete_ and therefore useless.
- As a way of bootstrapping fetchers that help fetch custom inputs, yes, but we don't have such a feature, so no.
- As a way of making sure that input declarations' evaluation doesn't require other inputs to be fetched, not really - we could just error out, or be clever about it.

So my conclusion thus far is that we could allow `inputs` to be proper expressions - not just data. After #7730 as mentioned.

### Alternatively, in the lock file

Another, broader solution is to write non-data expressions' outcomes to the lock file. That satisfies the requirement of no evaluation, but it might go out of sync, so it'd be up to you to commit the lock file if you care about others consuming your metadata. Tooling such as CI can help, without being as intrusive as custom templating. Externally consumable flakes should have CI anyway.

**Comment by jorsn** [comment]
I wrapped [my proposal](https://github.com/NixOS/nix/issues/4945#issuecomment-1877896238) in a [flake](https://github.com/jorsn/flakegen), to improve ease of use:
~~~
nix flake init -t github:jorsn/flakegen
nix run .#genflake flake.nix
~~~

Even if we might have a better solution in the future, this is usable *now*, if anyone is desperate to program flake inputs.

**Comment by sellout** [comment]
Rather than saying that a flake is “a subset of nix”, I think it’s generally helpful to describe a .nix file as containing either “a Nix expression” or “Nix data”[^1]. In practice it’s the distinction between whether the contents are evaluated before being acted on. A flake is not evaluated, but the handler for a flake will evaluate portions of it at different times.

[^1]: Fundamentally, a Nix expression is just Nix data where the data is a thunk of some type (`() -> a`), so it needs to be evaluated immediately (whereas, say, a module is Nix data of (roughly) type `{…} -> {config = …, options = …}`).

I think the richness of “Nix data” can be improved without compromising much, though. E.g.,
```nix
let
  name = "my-project";
  system = "x86_64-linux";
  nixpkgsRelease = "24.05";
in {
  description = name;

  outputs = inputs: {
    packages.${system}.${name} = …;
  };

  inputs.home-manager.url = "github:nix-community/home-manager/release-${nixpkgsRelease}";
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/release-${nixpkgsRelease}";
}
```
should only require “normalization”, not evaluation.

Since `outputs` is a function, users should already be encouraged to use tools (like `nix flake show` and `nix repl` with `:lf`) to explore a flake, rather than opening the file directly, so adding a bit more abstraction shouldn’t introduce a problem there.

Now, defining (and implementing) what is part of normalization is another question (e.g., should normalization execute `import`).

But I think regardless of normalization, the data/expression distinction is important and serves a general purpose (and should be documented somewhere so that questions like this can be directed there). Then, instead of talking about “what language” a flake (or some other .nix file) is in, we talk about the “type” of the data.

**Comment by ranomier** [comment]
Just force the output to be a different file and let flake.nix be mainly just the input file. 

I have this at the bottom of flake.nix. So much nicer.
```nix
  outputs = inArgs: import ./outputs.nix inArgs;
```

Then flake.nix is not a mixture of syntax.

**Comment by sellout** [comment]
@ranomier IMO, this issue isn’t about `outputs` in particular. For example, I just ran into another instance of this issue. I needed use my Cachix info somewhere else in my configuration, but I had to duplicate it because I’m not allowed to do

```nix
let
  cachix = {
    url = "https://….cachix.org";
    publicKey = "…";
  };
in {
  nixConfig = {
    extra-substituters = [cachix.url];
    extra-trusted-public-keys = [cachix.publicKey];
  };
  outputs = inputs: import ./outputs.nix {inherit cachix inputs;};
}
```

As I mentioned in my previous comment, I feel like any Nix file should at least be normalized[^2] first. Treating it like JSON[^3] discards most of the benefits.

[^2]: What “normalized” means needs to be specified (I don’t think there is anything between `parse` and `eval` in the current code, but I’m just a spectator). `let` is the main thing. If I can just extract a few constants like above, I’m pretty good. `nixpkgs.lib` wouldn’t be available, but should there be some partial evaluation of builtins? I would say it shouldn’t allow `rec`. If it allows `import`, that probably needs to be restricted (are circular imports already disallowed?). Does string interpolation happen? But maybe even restricted, it’s hard to eliminate all recursion loopholes – so then is it just a JSON structure, or do we just hope there are enough boundaries that people don’t stumble into them? (I would vote for the latter.)

[^3]: Even YAML has its weird anchors.

**Comment by ranomier** [comment]
@sellout Thank you for your time. I think i should clarify what my point is. It doesn't contradict with what we allow within the static flake part.

What i am trying to say is, whatever it is that is allowed within the flake part of that file, let's keep it within a file. Then we can say a flake.nix (maybe call it sources.nix or input.nix instead) is this particular format.

And let's keep the "output" also in a separate file.

From a UX and learning experience, it would be soooooo much easier to say things like.
- flake.nix is more static
- flake.nix doesn't run code
- In the output.nix you can run code
- etc ...

Additionally huge hurdle for me was and is to learn the custom language and being able to use more simpler language would be a really helpful thing.

---
2. yaml is not a good example^^ https://ruudvanasseldonk.com/2023/01/11/the-yaml-document-from-hell

**Comment by stepbrobd** [comment]
weird question

if i submit a patch that allows arbitrarily complex flake.nix file under a flag say `--trust-me-bro-it-will-terminate`, what are the chances of it getting accepted? i already have a [scratch PoC](https://github.com/stepbrobd/dotfiles/blob/523c487e11ac145495d98f68335bcfd24251aca7/modules/common/complex.patch) ready but definitely not ready for upstreaming

**Comment by roberth** [comment]
I've opened https://github.com/NixOS/rfcs/pull/193 to explore a solution to this.

**Comment by Malix-Labs** [comment]
I don't really know if NixOS/rfcs#193 could be considered as a "solution" to the fact that `flake.nix` is not a nix file but rather a subset of nix with a specific schema file

It can be a solution if it becomes **the only standard** to write flakes in, since they wouldn't mislead into thinking flakes are full nix files to begin with. _(which doesn't mean NixOS/rfcs#193 is desirable either)_

It seems like bikeshedding, but xkcd 927 is basically what I'm fearing :

![https://xkcd.com/927/](https://imgs.xkcd.com/comics/standards_2x.png)

**Comment by roberth** [comment]
Working towards a removal of flake.nix is one way to settle the question, but yeah, only in a literal interpretation of the question posed, and I do think it's only a compelling vision when combined with https://github.com/NixOS/rfcs/pull/194

**Comment by ruro** [comment]
Cross-posting from [my comment on the RFC](https://github.com/NixOS/rfcs/pull/193#discussion_r2597533442):

> I feel like this RFC should not just **assume** that the currently enforced status-quo limitation of "Flake input metadata can't contain arbitrarily complex, possibly non-terminating computations" must be kept as is, without justification. (...)
>
> For example, do we know if this restriction is actually useful / needed? I think, I read somewhere that FlakeHub is relying on the fact that flake metadata can be computed statically-ish (I've never used FlakeHub, so I am not sure if this is actually the case). Are there any more examples where this property is actually useful?
>
> If FlakeHub is the only (or one of the few) users of this restriction / guarantee, then maybe it should be implemented as a simple evaluation "depth" limiter on the side of FlakeHub, instead of in Flakes themselves?
>
> Or maybe it's enough to have a simple way to opt-out of this restriction? Something like
> ```nix
> {
>     dynamic = true;
>     inputs = {
>         # Arbitrary nix expressions are now allowed inside inputs,
>         # because `dynamic` was set to true. This flake might not have
>         # full functionality on services like FlakeHub, but this is fine.
>     };
> }
> ```

Honestly, the more time I spend thinking about this, the less I like the idea that forcing static metadata is the correct solution. Restricting `inputs` to be static metadata is only one possible solution and it feels really weird that everyone just assumes it to be a non-negotiable requirement.

As I've mentioned previously, here are some alternative options off the top of my head (roughly in order of personal preference, from best to worst):

1) Use "full" nix, but have an evaluation depth / complexity limit (enforce it only where it's actually needed, ie on FlakeHub or when loading a flake from an untrusted source or something).

2) Allow "full" nix, but also evaluate a static version of the metadata and save it in `flake.lock` (warn/error if static metadata doesn't match the locally evaluated results). Bonus points: you can also precompute and save some static information about the available outputs (which could be useful for searching / indexing purposes).

3) Use "static" subset of nix by default, but have a way to opt-out (for users that don't care about FlakeHub, etc).

4) Accept either `flake.nix` or `flake.toml`. Allow "full" nix inside `flake.nix`. That way, users that care about FlakeHub/etc will use `flake.toml` (or only the static subset of `flake.nix`), while everyone else can write arbitrary expressions in `flake.nix`.

P.S. I already asked this before, but does anyone have any concrete examples where this static metadata requirement is actually needed other than on FlakeHub?

**Comment by Malix-Labs** [comment]
Also see #15132, which is related about the last comment

https://github.com/NixOS/nix/pull/15132#issuecomment-4207461447 from @Malix-Labs :

> > Arguably it's more controlled than the alternative, which is that users have to cobble together input modifications using scripts that perform text replacements in flake.nix, or things along those lines.
> 
> This is a strong argument;
> People really wanting to compute their flakes inputs are just doing it regardless, in an unsafer way.

**Comment by Malix-Labs** [comment]
https://github.com/NixOS/nix/issues/4945#issuecomment-868572873 from @edolstra :
> This should be documented in the flake manpage.

And also in https://nix.dev/manual/nix/latest/command-ref/new-cli/nix3-flake.html#flake-format and https://nix.dev/manual/nix/2.34/command-ref/new-cli/nix3-flake.html#flake-inputs

Currently, there is no clear mentions of the flake inputs limitations raised in this issue

**Comment by Malix-Labs** [comment]
I just discovered https://github.com/denful/flake-file

@roberth is right in https://github.com/NixOS/nix/pull/15132#issuecomment-4207131122 :

> I'm not convinced that Flakes users need to be patronized in this way, or that IFD is particularly problematic here.
Arguably it's more controlled than the alternative, which is that users have to cobble together input modifications using scripts that perform text replacements in flake.nix, or things along those lines.

I am now completely convinced that nix should indeed allow non-trivial expressions in flakes (#15132);
To which extends is the question, and the current extends is unsatisfactory

I am also convinced that if NixOS/nix#15132 was to be rejected, then NixOS/rfcs#193 (or another non-turing configuration language) AND deprecating `flake.nix` would be the only legitimate alternative to the concept of nix flakes;

Either flakes inputs are defined in a non-turing configuration language, either they are defined in Nix

The current state of flake being a loosely defined subset of nix is objectively wrong because of this indecision

**Comment by mightyiam** [comment]
I'm a user of flake-file because it allows me to declare inputs in different files. If flake inputs could be evaluated dynamically that would obviate the need to run a command that generates the `flake.nix` and that would make the workflow nicer and probably make flake-file obsolete, in favor of some built-in flake-parts support for declaring flake inputs.

Having said that, I would support a path similar to Lix, where flake features are frozen until the flakes are well defined and stable or whatever else ends up being their fate.

**Comment by vaxerski** [comment]
I have to use flake-file to avoid the absolute ugliness of flake.nix for my needs.

I develop hyprland, and I need to be dogfooding my own software on the absolute latest revs. 

Now, problem is, when you have like 10 apps and 10 libraries, it turns into this fiesta of:

```nix
app = { url = "..."; inputs.lib1.follows = "..."; inputs.lib2.follows = "..."; inputs.lib3.follows = "..."; inputs.lib4.follows = "..."; };
```

which is a) an eyesore and b) a maintenance nightmare, because if any dep gets added, I need to manually keep track of it and add it.

With flake-file, I can actually generate those automatically.

The basic flake.nix seems awfully restricted.

**Comment by Malix-Labs** [comment]
See https://github.com/NixOS/nix/pull/15132#issuecomment-4207131122 from @roberth, which is an instance of what Vax is doing

**Comment by mightyiam** [comment]
While this is offtopic, it's a short point and might serve @vaxerski and those who read his comment on the proliferation and handling of transitive inputs.

If a flake input can be consumed not via its flake outputs but instead by importing a nix file by path from it or perhaps by using it as a source, then its transitive inputs can be avoided by setting `flake = false;` (to be clear, the attr path is `inputs.<name>.flake`). This opts out of evaluating the inputs `flake.nix` and `flake.lock`. I do this whenever possible.
